api/openapi-spec.yaml

```diff
@@ -0,0 +1,473 @@
+openapi: 3.0.1
+info:
+  title: Event Repository REST API
+  description: Event Repository REST API for retrieving the event information
+  contact: {}
+  version: 0.0.1
+tags:
+- name: event-resource
+  description: The Event Resource API for getting single event information
+- name: search-resource
+  description: The search api provides methods for querying up/down stream given an
+    event id.
+- name: events-resource
+  description: The Events Resource API for getting all events information
+paths:
+  /events:
+    get:
+      tags:
+      - events-resource
+      summary: To get all events information
+      operationId: getEventsUsingGET
+      parameters:
+      - name: pageNo
+        in: query
+        description: "Page to display if results span across multiple pages."
+        schema:
+          type: integer
+          format: int32
+          default: 1
+      - name: pageSize
+        in: query
+        description: "The number of events to be displayed per page."
+        schema:
+          type: integer
+          format: int32
+          default: 500
+      - name: pageStartItem
+        in: query
+        description: "Intended to skip few items at the start of result, Should\
+          \ be used only if `pageNo=1` ie. first page of results."
+        schema:
+          type: integer
+          format: int32
+          default: 1
+      - name: shallow
+        in: query
+        description: "Determines if external ER's should be used to compile\
+          \ the results of query. Use `false` to use External ER's."
+        schema:
+          type: boolean
+          default: false
+      - name: lazy
+        in: query
+        description: "If lazy is `true`, it implies that when the events limit\
+          \ is reached according to pazesize no additional request is performed and the search will stop."
+        schema:
+          type: boolean
+          default: false
+      - name: readable
+        in: query
+        description: |
+          Determines if event time should be in milliseconds, or in Human readable time format. Ex:
+
+          `false` 1499076742982
+
+          `true`  2018-10-31T13:36:00.824Z.
+        schema:
+          type: boolean
+          default: false
+      - name: params
+        in: query
+        description: |
+          To search for specific events or artifacts, filtering with parameters is supported.
+
+          **Syntax:**
+
+          `<resource>?key[.key ...]<FC>value[&key[.key ...]<FC>value ...]`
+
+          <FC> is one of the Filter comparators described below. To traverse into nested structures and filter on their keys, namespacing with
+          `.` (dot) is used.
+
+          ```
+          =  - equal to
+          >  - greater than
+          <  - less than
+          >= - greater than or equal to
+          <= - less than or equal to
+          != - not equal to
```

Comment:
The main use I seen for this is if you want to select events within a data range. E.g. if you want to select the event sent during a month or a week.

If we want to get a small spec we could skip the rest and add it back later if needed.

Reply:
Okay. Such interactive use is better dealt with with a proper web frontend. Also, the current spec with additional operators isn't exactly easy to use either since >, !, < and so on must be percent-encoded.

Reply:
Several examples in the /events endpoint docs still reference operators other than `=`.

Reply:
Also, there are references to type conversions. Are they relevant for equality comparisons?

Reply:
Our current implementation treats all incoming values as strings but in the database we have e.g. time values stored as doubles(?) thus to make a match we have to use type conversions.

Other issues should be fixed in next commit

Reply:
Okay, but surely clients can't be expected to track the internal representation of fields in the ER implementation and adapt their queries based on that? The understanding that a floating-point field in the database may require type conversions of operands must reside on the server side, hidden from clients of the API. Same thing with bool fields; shouldn't the ER be able to do the right thing when given `someBoolField=true`? And not require `bool(someBoolField)=true`?

After excluding less/greater than operators from the spec I see no reason to have type conversions.

Reply:
Fixed in [93399a9](https://github.com/eiffel-community/eiffel-event-repository/pull/12/commits/93399a979e8bf4b7dfdb43ad1140cd33fef92a9e)

Reply:
@magnusbaeck OK?

Reply:
Yes, thanks!